### PR TITLE
[FEQ]Jim/FEQ-2041/add semantic release

### PR DESCRIPTION
**Pull request by jim-deriv**
## Changes:
- Added workflow for triggering semantic release on push to `master`
- Added semantic release configurations

### CODE REVIEW

File: release.config.cjs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+const transform = require('./release.utils.cjs');
+/**
+ * @type {import('semantic-release').GlobalConfig}
+ */
+module.exports = {
+    branches: ['master'],
```
Comment (jim-deriv) [inline release.config.cjs:6]: Not sure if we'll need `pre-release` branches. I'll add them incase we need them.
```suggestion
    branches: ['master'],
```

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      id-token: write
+    steps:
+      - name: Checkout
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11
+        with:
+          fetch-depth: 0
+      - name: Setup Node.js
+        uses: actions/setup-node@b39b52d1213e96004bfcb1c61a8a6fa8ab84f3e8
+        with:
+          node-version: 21.x
```
Comment (hasan-deriv) [inline .github/workflows/release.yml:23]: why it is 21?

Comment (jim-deriv) [reply .github/workflows/release.yml:23]: Because it's the `latest` and it's stable. Do you have any concerns with it? 🤔

Comment (shahzaib-deriv) [reply .github/workflows/release.yml:23]: @jim-deriv I suggest sticking to Node 20 LTS as of now v21 is not getting any updates other than security updates which are going to be ended in a month

ref: https://endoflife.date/nodejs

**Review by shahzaib-deriv — APPROVED** [review]

**Review by hasan-deriv — APPROVED** [review]

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+name: Release
+on:
+    push:
+        branches:
+            - master
```
Comment (niloofar-deriv) [inline .github/workflows/release.yml:5]:
```suggestion
            - master
            - beta
      - next
      - '^([0-9]|[1-9][0-9]*)\.([0-9]|[1-9][0-9]*)\.([0-9]|[1-9][0-9]*|x)'
```

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+jobs:
+    release:
+        name: Build and Release
+        runs-on: ubuntu-latest
+        permissions:
+            contents: write
```
Comment (niloofar-deriv) [inline .github/workflows/release.yml:11]: Do we need write permission for this? 

I guess read would be ok
```suggestion
            contents: read
```

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            issues: write
+            pull-requests: write
```
Comment (niloofar-deriv) [inline .github/workflows/release.yml:13]: Do we need all of these permissions? 
AFAIK we just need contents permission 🤔

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        steps:
+            - name: Checkout
+              uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11
+              with:
+                  fetch-depth: 0
+            - name: Setup Node.js
+              uses: actions/setup-node@b39b52d1213e96004bfcb1c61a8a6fa8ab84f3e8
+              with:
+                  node-version: 20.x
+            - name: Install Dependencies
```
Comment (niloofar-deriv) [inline .github/workflows/release.yml:24]: Can we replace this step with the one in the shared action 🤔 
https://github.com/niloofar-deriv/shared-actions/tree/master/.github/actions

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+              run: npm clean-install
+            - name: >-
+                  Verify The Integrity of Provenance Attestations and Registry
+                  Signatures for Installed Dependencies
+              run: npm audit signatures
+            - name: Release
+              env:
```
Comment (niloofar-deriv) [inline .github/workflows/release.yml:31]:
```suggestion
              env:
              CI: true
```

Comment (niloofar-deriv) [reply .github/workflows/release.yml:31]: I guess we need this as well 🤔

File: .github/workflows/release.yml (hunk 3 above)
Comment (jim-deriv) [reply .github/workflows/release.yml:5]: @niloofar-deriv, as I mentioned in [this](https://github.com/deriv-com/deriv-api-hooks/pull/26#discussion_r1569979549) comment, as far as I know, we only release from the `master` branch. We don't release from any other branch. When needed, it'll be added.

File: .github/workflows/release.yml (hunk 4 above)
Comment (jim-deriv) [reply .github/workflows/release.yml:11]: This is needed to be able to publish a GitHub release

File: .github/workflows/release.yml (hunk 5 above)
Comment (jim-deriv) [reply .github/workflows/release.yml:13]: We need permissions for pull requests and issues to be able to write [these](https://github.com/deriv-com/deriv-utils/pull/38#issuecomment-2044675729) comments

File: .github/workflows/release.yml (hunk 7 above)
Comment (jim-deriv) [reply .github/workflows/release.yml:31]: As far as I remember, this is set to true by default

File: .github/workflows/release.yml (hunk 6 above)
Comment (jim-deriv) [reply .github/workflows/release.yml:24]: Alright, I'll replace it with that.

**Review by thisyahlen-deriv — APPROVED** [review]

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+              uses: actions/setup-node@b39b52d1213e96004bfcb1c61a8a6fa8ab84f3e8
+              with:
+                  node-version: 20.x
+            - name: Install Dependencies
+              run: npm clean-install
+            - name: >-
+                  Verify The Integrity of Provenance Attestations and Registry
+                  Signatures for Installed Dependencies
+              run: npm audit signatures
+            - name: Release
```
Comment (shayan-deriv) [inline .github/workflows/release.yml:30]:
```suggestion
            - name: Release
               if: success()
```
don't we need this condition checking? 🤔

File: release.config.cjs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                    mergeCorrespondence: ['id', 'source'],
+                },
+                writerOpts: {
+                    transform: transform,
+                },
+            },
+        ],
+        '@semantic-release/changelog',
+        '@semantic-release/npm',
+        '@semantic-release/github',
```
Comment (shayan-deriv) [inline release.config.cjs:42]: you need to add this to your package.json devDependencies as well

File: .github/workflows/release.yml (hunk 8 above)
Comment (niloofar-deriv) [reply .github/workflows/release.yml:30]: it has this check by default

Comment (jim-deriv) [reply .github/workflows/release.yml:30]: I don't think we need it @shayan-deriv, as far as I remember, when  any of the of the steps fails, the rest are skipped

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -0,0 +1,35 @@
+name: Release
```
Comment (niloofar-deriv) [inline .github/workflows/release.yml:1]: That would be great if we could move this semantic release workflow to the shared action repo cause we are repeating same things in different projects

**Review by niloofar-deriv — APPROVED** [review]

File: release.config.cjs (hunk 9 above)
Comment (jim-deriv) [reply release.config.cjs:42]: Yes @shayan-deriv, any `dependencies` that don't come bundled with `semantic-release` need to be added to the `package.json`

File: .github/workflows/release.yml (hunk 10 above)
Comment (jim-deriv) [reply .github/workflows/release.yml:1]: Yes @niloofar-deriv, if you create a PR, I'll be the first one to approve it 😁

**Review by thisyahlen-deriv — APPROVED** [review]

File: .github/workflows/release.yml (hunk 10 above)
Comment (niloofar-deriv) [reply .github/workflows/release.yml:1]: Haha sure @jim-deriv 😂

**Pull request merged**